feat: Add LIP-33: Community Staking Module v3 and Curated Module v2 (#108)
lidofinance/lido-improvement-proposals: LIPS/lip-33.md

Reviewer: maybe we can just add smth like a dedicated "Deployment Model" or "Migration" subsection early in the spec that states:

- CSM v3 with 0x02 WC = new contract deployment, new StakingRouter module ID
- Existing 0x01 CSM = upgraded in-place with v3 contract code but stays 0x01
- Whether operators can/should migrate validators from 0x01 to 0x02 instance, and how

Reviewer: also can be a backward compat section, idk

---
lip: 33
title: Community Staking Module v3 and Curated Module v2
status: Proposed
author: Dmitry Gusakov (@dgusakov), Sergey Khomutinin (@skhomuti), Dmitry Chernukhin (@madlabman), Vladimir Gorkavenko (@vgorkavenko)

> discussions-to: https://research.lido.fi/t/community-staking-module/5917

Reviewer: discussions about CMv2 also there?

> discussions-to: https://research.lido.fi/t/community-staking-module/5917

Reviewer: discussions about CMv2 also there?

created: 2026-03-18
updated: 2026-03-18
---

# LIP-33. Community Staking Module v3 and Curated Module v2

## Simple Summary

> Community Staking Module v3 (CSM v3) is a technical upgrade to the existing [CSM v2](https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-29.md). It features native support of `0x02` validator withdrawal credentials (assuming operations when each `0x02` validator is fully deposited to `2048 ETH`) and built-in Node Operator reward splitters. Additionally, several technical improvements have been made, including improved penalty handling, an optimized reward-claim mechanism, and more granular role-based access controls for managing Node Operator type parameters.

Reviewer: ```suggestion
Community Staking Module v3 (CSM v3) is a technical upgrade to the existing [CSM v2](https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-29.md). It features native support of `0x02` validator withdrawal credentials (assuming operations when each `0x02` validator is fully deposited to `2048 ETH`) however under separate instance, not the existing one, and built-in Node Operator reward splitters. Additionally, several technical improvements have been made, including improved penalty handling, an optimized reward-claim mechanism, and more granular role-based access controls for managing Node Operator type parameters.
```

> Community Staking Module v3 (CSM v3) is a technical upgrade to the existing [CSM v2](https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-29.md). It features native support of `0x02` validator withdrawal credentials (assuming operations when each `0x02` validator is fully deposited to `2048 ETH`) and built-in Node Operator reward splitters. Additionally, several technical improvements have been made, including improved penalty handling, an optimized reward-claim mechanism, and more granular role-based access controls for managing Node Operator type parameters.

Reviewer: ```suggestion
Community Staking Module v3 (CSM v3) is a technical upgrade to the existing [CSM v2](https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-29.md). It features native support of `0x02` validator withdrawal credentials (assuming operations when each `0x02` validator is fully deposited to `2048 ETH`) however under separate instance, not the existing one, and built-in Node Operator reward splitters. Additionally, several technical improvements have been made, including improved penalty handling, an optimized reward-claim mechanism, and more granular role-based access controls for managing Node Operator type parameters.
```

> Curated Module v2 (CM v2) is a new chapter for the Lido protocol. Based on the solid codebase of CSM, it inherits new concepts introduced in CSM v3, like support for `0x02` validators, integration features, and many more. A robust bonding mechanism allows CM v2 to streamline Node Operator management, reduce operational complexity, and increase protocol security. With CM v2, the curated part of the Lido Validator set will benefit from all the advantages CSM has to offer.

Reviewer: ```suggestion
Curated Module v2 (CM v2) is a new chapter for the Lido protocol. Based on the solid codebase of CSM, it inherits new concepts introduced in CSM v3, like support for `0x02` validators, integration features, and many more. A robust bonding mechanism allows CM v2 to streamline Node Operator management, reduce operational complexity, and increase protocol security. With CM v2, the curated part of the Lido Validator set will benefit from all the advantages CSM has to offer. Currently existing Curated Module v1 will undergo a sunset process by migrating stake into CM v2 throughout the coordinated migration campaign.
```

> ## Motivation

Reviewer: I love how it shaped now, but would shift a bit from marketing/user-facing perspective towards systemic protocol one:
- How much stake sits in the current Curated Module v1? (~200k+ validators?) What fraction of total Lido TVL?
= What's the concrete benefit of migrating that stake to 0x02? (e.g., consolidation from ~200k validators to ~N MaxEB validators reduces attestation load by X%, saves Y ETH/year in gas for exits)
- What happens if we don't do this? L"if the update is not delivered, X breaks.", e.g. what breaks or degrades if CSM stays on 0x01 and CM stays on v1?

would probably rely on these materials as a strategic roadmap that this LIP makes to happen:
- https://blog.lido.fi/lidos-roadmap-to-pectra-delivering-validator-consolidations-in-the-protocol/
- https://blog.lido.fi/lidos-roadmap-to-pectra-navigating-complexity/
- https://blog.lido.fi/lido-and-pectra-ensuring-compatibility/

> Over the past 1.5 years, CSM has proven to be the most reliable and scalable staking module in the Lido protocol. Fundamental concepts such as validator bonds, automated performance management, and tailored Node Operator types have meaningfully improved the Lido protocol's resilience and decentralization. Today, we introduce the next chapter of the staking modules in the Lido protocol: the CSM upgrade, featuring support for the much-anticipated 0x02 validator withdrawal credentials and more. `0x02` validator WCs are essential for the Ethereum network's scalability and future upgrades. Bundled with a few more technical improvements, CSM v3 will make CSM even more user-friendly and reliable.

Reviewer: it's not clear that the currently deployed instance won't support 0x02 and we rather expect to deploy a separate one for that exact purpose

> In addition to CSM upgrade the time has come for the Curated Module to evolve. Being almost unchanged since the Lido v2 release, the Curated module is meaningfully outdated by now, so we need to introduce a new version of the Curated module to allow for future evolution, as described in the [forum post](https://research.lido.fi/t/future-of-the-curated-module-cmv2-landscape/10929). Curated Module v2 will be a new deployment and will assume migration from the old version via validator consolidations. Using CSM v3 as a base, Curated Module v2 features several unique mechanics such as Meta Operators Registry and Weighted Stake Allocation. These features lay a solid foundation for the future evolution of the Curated Module that encompasses the Validation Market. While the market is not included in the scope of the current release, it will be added in the future. For now, the main feature that will be enabled by the Curated Module v2 is support for `0x02` validator WCs. This will allow the Lido protocol to migrate the majority of its protocol stake to 0x02 validators, significantly boosting 0x02 WC adoption at the Ethereum network level.

Reviewer: ```suggestion
In addition to CSM upgrade the time has come for the Curated Module to evolve. Being almost unchanged since the Lido v2 release, the Curated module design doesn't fit into the modern vision, so we need to introduce a new version of the Curated module to allow for future evolution, as described in the [forum post](https://research.lido.fi/t/future-of-the-curated-module-cmv2-landscape/10929). Curated Module v2 will be a new deployment and will assume migration from the old version via validator consolidations. Using CSM v3 as a base, Curated Module v2 features several unique mechanics such as Meta Operators Registry and Weighted Stake Allocation. These features lay a solid foundation for the future evolution of the Curated Module that encompasses the Validation Market. While the market is not included in the scope of the current release, it will be added in the future. For now, the main feature that will be enabled by the Curated Module v2 is support for `0x02` validator WCs. This will allow the Lido protocol to migrate the majority of its protocol stake to 0x02 validators, significantly boosting 0x02 WC adoption at the Ethereum network level.
```

- The EL rewards stealing penalty mechanism is extended to a general penalty.

#### CM-only contracts

##### `CuratedModule.sol`

*CuratedModule on the scheme* 

> New in CM v2

`CuratedModule.sol` is a core module contract conforming to the `IStakingModule` and `IStakingModuleV2` interfaces. The contract inherits from `BaseModule.sol` (which contains all common parts with CSM) to provide extended functionality while maintaining consistency with the original CSM. It stores information about Node Operators and deposit data (DD). This contract is responsible for all interactions with the `StakingRouter`, namely, the DD management and some of the Node Operator's parameters. Node Operators manage their validator keys and other parameters that they can modify through this contract.

**Changes compared to CSM v3:**
- The management system for the NodeOperators address has been changed to allow custom role members to set both rewardAddress and managerAddress in an emergency.
- Stake distribution is done via a ["greedy" weighted stake allocation strategy](#Allocation-Strategy).

##### `CuratedGate.sol`

*CuratedGates on the scheme*

> New in CM v2

> `CuratedGate.sol` is a supplementary contract that enables Node Operator creation for the vetted addresses, which serves as an entry point to [`CuratedModule.sol`](#CuratedMosulesol). Alongside Node Operator creation, a contract can assign a custom Node Operator type (bondCurveId) in `Accounting.sol`. Deployed using `MerkleGateFactory.sol` to allow the addition of the new instances later without additional code security audits. The list of curated participants is individually upgradable for each instance of the [`CuratedGate.sol`](#CuratedGatesol) using a dedicated EasyTrack factory.

Reviewer: ```suggestion
`CuratedGate.sol` is a supplementary contract that enables Node Operator creation for the vetted addresses, which serves as an entry point to [`CuratedModule.sol`](#CuratedModulesol). Alongside Node Operator creation, a contract can assign a custom Node Operator type (bondCurveId) in `Accounting.sol`. Deployed using `MerkleGateFactory.sol` to allow the addition of the new instances later without additional code security audits. The list of curated participants is individually upgradable for each instance of the [`CuratedGate.sol`](#CuratedGatesol) using a dedicated EasyTrack factory.
```

**Changes in CM v2:**
- The bot can now also serve CM v2. There will be separate instances for CSM and CM.

##### `CM Oracle`

> Changed in CM v2

`CM Oracle` (also known as CM Performance Oracle) is a module in the common Lido on Ethereum (LoE) Oracle set. It is operated by the existing Oracle set alongside [Accounting Oracle](https://docs.lido.fi/contracts/accounting-oracle), [Validator Exit Bus Oracle](https://docs.lido.fi/contracts/validators-exit-bus-oracle), and CSM Performance Oracle. It is responsible for calculating the CM v2 Node Operators' reward distribution and strike assignment based on their performance on the CL. `CM Oracle` uses the same code as [`CSM Oracle`](https://hackmd.io/@lido/csm-v3-spec#CSM-Oracle) does.

##### `CMC`

> New in CM v2

`CMC` (also known as CM Committee Multisig) is an off-chain committee responsible for overseeing CM v2 and managing several flows and operations within it. Somewhat similar to [Community Staking Module Committee](https://research.lido.fi/t/community-staking-module-committee/8333) Multisig, but with the extended permissions.

### Fundamental concepts of CM

#### Meta Operators Registry

> [`CuratedModule.sol`](#CuratedMosulesol) uses the [`MetaRegistry.sol`](#MetaRegistrysol) contract to fetch information about Node Operators' stake and allocation weights. [`MetaRegistry.sol`](#MetaRegistrysol) has a permissioned method (expected to be called by [CMC](#CMC)) to add information about `OperatorGroup`. `OperatorGroup` has the following form:

Reviewer: ```suggestion
[`CuratedModule.sol`](#CuratedModulesol) uses the [`MetaRegistry.sol`](#MetaRegistrysol) contract to fetch information about Node Operators' stake and allocation weights. [`MetaRegistry.sol`](#MetaRegistrysol) has a permissioned method (expected to be called by [CMC](#CMC)) to add information about `OperatorGroup`. `OperatorGroup` has the following form:
```

Once uploaded, deposit data becomes available for allocation of the initial `32 ETH` deposits. To allocate initial deposits to the CM v2 Node Operators, the `StakingRouter` calls the `obtainDepositData(depositsCount)` method to retrieve the next `depositsCount` depositable keys, as determined by the stake [allocation strategy](#Allocation-Strategy).

##### Top-ups for `0x02` keys

![Top-up deposit Curated](./assets/lip-33/top_up_deposit_curated.png)

`0x02` validators get deposits in two phases:
- [Initial `32 ETH` deposit](#Initial-32-ETH-deposits).
- Top-ups up to `2048 ETH`.

The top-up phase involves a call from the `StakingRouter` to [`CuratedModule.sol`](#CuratedModulesol), providing information about the keys planned for top-up and top-up limits calculated using CL proofs. [`CuratedModule.sol`](#CuratedModulesol) calculates the stake allocation using the `maxDepositAmount` provided and allocates the stake among the provided keys, respecting the calculated allocation and the `topUpLimits` supplied by the `StakingRouter`. [`CuratedModule.sol`](#CuratedModulesol) also verifies that the keys provided belong to the corresponding Node Operators for security reasons.

> ##### Deposits order and priority

Reviewer: ❤️

- **Consolidations from the legacy CM v1.** This is the third source of balance increase for the validator. In the same manner as a top-up deposit, it can only be made (both from the Lido protocol's security perspective and under Ethereum network rules) to a validator with an assigned index. Moreover, the target validator (the one that will receive the balance after consolidation) should also be active on the Ethereum network.

Given the above-mentioned, the following guidelines for deposits is used in CM v2:
- **Initial deposits have the highest priority.** Whenever there are keys that can receive initial 32 ETH deposits, buffered ETH is used for them.
- It is recommended to use the earliest deposited keys as targets for consolidations.
- Top-up deposits should not clash with consolidations and respect the current and pending validator balance when submitted.

The actual process of deposits for CM v2 looks as follows:
- After module activation, the first phase consists only of initial deposits to all available keys, with consolidations and top-ups disabled.

> - Once there is a significant amount of activated keys in CM v2, the consolidation process starts, and top-up deposits are enabled.

Reviewer: "significant amount of activated keys" seems underspecified tbh

##### New issue

In the updated versions of CSM and CM v2, we compare the withdrawal balance to the `max_reported_validator_balance` that is permissionlesly delivered by prover-bot to correctly account for the top-ups that were not applied on CL yet.

Since top-ups are now the essential part of the deposit flow in the protocol, it is possible for `0x02` modules that:
1. The deposit queue on the Ethereum network is long enough
2. A top-up is made to the validator
3. This validator exits soon after the deposit and gets withdrawn before the deposit is applied on CL
4. The withdrawal event is not proven on-chain for some reason (prover-bot failure)
5. Top-up is applied on CL and gets withdrawn
6. If the amount of the top-up that was just withdrawn fits into `[max_reported_validator_balance * (100% - MIN_WITHDRAWAL_RATIO), max_reported_validator_balance]` Node Operator's bond will be penalized for `max_reported_validator_balance - topup_amount`

> Alternatively, the simmilar case can happen when the validator should have been penalized because `exit_balance < max_reported_validator_balance`, but there was a top-up with the `amount > max_reported_validator_balance`, and this top-up withdrawal is used as a withdrawal proof. In this the losses are for the protocol and not Node Operator since the penalty that should have been applied is not applied.

Reviewer: ```suggestion
Alternatively, the similar case can happen when the validator should have been penalized because `exit_balance < max_reported_validator_balance`, but there was a top-up with the `amount > max_reported_validator_balance`, and this top-up withdrawal is used as a withdrawal proof. In this the losses are for the protocol and not Node Operator since the penalty that should have been applied is not applied.
```
